experimental/InvariantTheory/types.jl

```diff
@@ -69,3 +69,55 @@ struct InvRingBasisIterator{InvRingT, IteratorT, PolyElemT, MatrixT}
   monomials_collected::Vector{PolyElemT}
   kernel::MatrixT # used iff reynolds == false
 end
+
+abstract type VectorSpaceIterator{FieldT, IteratorT, ElemT} end
+
+# This takes a basis of a vector space as an iterator and then "iterates" the
+# space in three "phases":
+# 1) iterate the basis using basis_iterator, so return one basis element at
+#    a time
+# 2) iterate all possible sums of basis elements
+# 3) return random linear combinations of the basis elements (with integer
+#    coefficients bounded by rand_bound)
+#
+# We collect the basis while iterating it, so that multiple iteration over the
+# same VectorSpaceIterator do not need to recompute it. basis_iterator_state
+# must always be a state of basis_iterator, so that
+# iterate(basis_iterator, basis_iterator_state) returns the next element in the
+# basis coming after basis_collected[end].
+mutable struct VectorSpaceIteratorRand{FieldT, IteratorT, ElemT} <: VectorSpaceIterator{FieldT, IteratorT, ElemT}
+  field::FieldT
+  basis_iterator::IteratorT
+  basis_collected::Vector{ElemT}
+  basis_iterator_state::Any # I don't know the type of this and I don't think there
+                       # is a "type-stable" way of finding it out
+  rand_bound::Int
+
+  function VectorSpaceIteratorRand(K::FieldT, basis_iterator::IteratorT, bound::Int = 10^5) where {FieldT, IteratorT}
+    VSI = new{FieldT, IteratorT, eltype(basis_iterator)}()
+    VSI.field = K
+    VSI.basis_iterator = basis_iterator
+    VSI.basis_collected = eltype(basis_iterator)[]
+    VSI.rand_bound = bound
+    return VSI
+  end
+end
+
+# The same things as for VectorSpaceIteratorRand apply besides that in "phase 3"
+# all elements of the vector space are iterated in deterministic order (it is
+# supposed to be finite after all).
+mutable struct VectorSpaceIteratorFiniteField{FieldT, IteratorT, ElemT} <: VectorSpaceIterator{FieldT, IteratorT, ElemT}
+  field::FieldT
+  basis_iterator::IteratorT
+  basis_collected::Vector{ElemT}
+  basis_iterator_state # I don't know the type of this and I don't think there
```

Comment:
Of course this should also get `Any` (usually a good idea to look around if a given code review remark might also apply in other places :-) )
```suggestion
  basis_iterator_state::Any # I don't know the type of this and I don't think there
```